### PR TITLE
Merge upstream HTTP changes into mail branch

### PR DESCRIPTION
**Pull request by m1cr0man**
In order to proceed with setting up the mail stack, specifically mailman, I had to update the nixpkgs version I was working off. This meant a few things:
- The rbacme.nix module was merged to upstream and so I could remove it
- Privatebin is now working off master too
- services.httpd.virtualHosts was changed from a list to a dict so I had to rewrite a lot of our stuff to support this

I am merging this into mail branch so that the diff is easier to read. I need this merged soon ish. Mail is nearly ready for merge to master (as in, I might have it done tonight)

### CODE REVIEW

File: common/variables.nix [hunk 1]
```diff
@@ -4,13 +4,36 @@ rec {
...
 
+  # This maps some domains to other domains
+  # usually due to broken DNS entries
+  # Allows certs to be correctly generated
+  brokenDomains = {
+    "djbdns.now.ie" = "djbdns.now.ie";
+    "romana.now.ie" = "djbdns.now.ie";
+    "www.luxgaa.lu" = "www.luxgaa.lu";
+    "www.iahpc.ie" = "www.iahpc.ie";
+    "techweek.dcu.ie" = "techweek.dcu.ie";
```
Comment (butlerx) [inline common/variables.nix:17]: mind explaining this a bit?

File: services/httpd/default.nix [hunk 2]
```diff
@@ -10,9 +10,7 @@ let
   # Everything else is promoted to HTTPS
   acmeVhost = {
     inherit adminAddr;
-    hostName = tld;
     serverAliases = ["*"];
-    listen = [{ port = 80; }];
```
Comment (butlerx) [inline services/httpd/default.nix:15]: this covered by default or what?

File: services/httpd/default.nix [hunk 3]
```diff
@@ -28,9 +26,9 @@ let
     documentRoot = "${common.webtreeDir}/redbrick/htdocs";
   in {
     inherit adminAddr documentRoot;
-    hostName = tld;
-    listen = [{ port = 443; }];
```
Comment (butlerx) [inline services/httpd/default.nix:32]: same question as above

File: services/httpd/shared.nix [hunk 4]
```diff
@@ -5,52 +5,43 @@ let
...
-    enableSSL = true;
+  vhost = {user, group, documentRoot, serverAliases ? [],
+            extraConfig ? "", wwwRedirect ? false}: {
+    inherit documentRoot serverAliases;
+    adminAddr = if user == "wwwrun" then adminAddr else "${user}@${tld}";
+    onlySSL = true;
     extraConfig = ''
       <Directory "${documentRoot}">
-        <FilesMatch \.php\d*$>
+        <FilesMatch \.php\d*$>hostName
```
Comment (butlerx) [inline services/httpd/shared.nix]: whats hostname doing here?

File: services/httpd/vhosts.nix [hunk 5]
```diff
@@ -299,234 +256,203 @@ in (if (config.redbrick.skipVhosts) then [] else ([
...
-  (vhostRedirect "helpdeskexam.${tld}" "https://md.redbrick.dcu.ie/s/SJzip7F9X#")
-  (vhostRedirect "hoodies.${tld}" "https://redbrickdcu.typeform.com/to/Q4uIzR")
-  (vhostRedirect "parlour.${tld}" "http://www.songsfromtheparlour.com")
-  (vhostRedirect "radio.theinternets.be" "http://radio.redbrick.dcu.ie")
-  (vhostRedirect "techweek.${tld}" "https://techweek.dcu.ie")
-  (vhostRedirect "tickets.${tld}" "https://dcusu.ticketsolve.com/shows/873599383/events/128190598")
-  (vhostRedirect "ubuntu.${tld}" "https://wiki.redbrick.dcu.ie/mw/RedBrick_Ubuntu")
-] ++ userVhosts))
+  };
+  "dcufm.${tld}" = vhostProxy "http://136.206.15.74";
```
Comment (butlerx) [inline services/httpd/vhosts.nix:427]: we seem to be missing all the metrics services. did you rebase of master?

File: common/variables.nix (hunk 1 above)
Comment (m1cr0man) [reply common/variables.nix:17]: Part of the script for generating certs will break down domains to their tld. So for example, www.iahpc.ie will become iahpc.ie. In most cases, this is totally fine, but for these few sites their tld's do not point at redbrick and thus we can't generate HTTP-validated certs for them. This list allows us to work around this issue. I might add a more descriptive comment.

File: services/httpd/default.nix (hunk 2 above)
Comment (m1cr0man) [reply services/httpd/default.nix:15]: if you don't specify any of the SSL options it only listens on port 80. the listen option itself got deprecated.

File: services/httpd/default.nix (hunk 3 above)
Comment (m1cr0man) [reply services/httpd/default.nix:32]: Yeah onlySSL covers this, as mentioned above

File: services/httpd/shared.nix (hunk 4 above)
Comment (m1cr0man) [reply services/httpd/shared.nix]: typo

File: services/httpd/vhosts.nix (hunk 5 above)
Comment (m1cr0man) [reply services/httpd/vhosts.nix:427]: Not yet, I won't be doing that until I get mail ready to merge. I'll cover those then

Comment (m1cr0man) [reply services/httpd/vhosts.nix:427]: Yeah bare in mind this merges into the `mail` branch

**Review by butlerx — APPROVED** [review]

**Review by theycallmemac — APPROVED** [review]

**Pull request merged**